### PR TITLE
add volume properties changed event

### PR DESCRIPTION
**Pull request by NNEU-1**

### CODE REVIEW

File: src/dbus_mpris.rs [hunk 1]
```diff
@@ -438,42 +439,54 @@ async fn create_dbus_server(
...
             }
             PlayerEvent::Paused { .. } => {
-                (last_track_id, Some(PlaybackStatus::Paused))
+                (last_track_id, Some(PlaybackStatus::Paused), last_volume)
             }
-            _ => (last_track_id, last_playback_status),
+            _ => (last_track_id, last_playback_status, last_volume),
         };
 
         // if playback_status or track_id have changed, emit a PropertiesChanged signal
```
Comment (eladyn) [inline src/dbus_mpris.rs]: nitpick: this comment should be updated

File: src/dbus_mpris.rs [hunk 2]
```diff
@@ -438,42 +439,54 @@ async fn create_dbus_server(
...
+                    changed_properties.insert("Volume".to_owned(), Variant(Box::new(player_volume.to_owned())));
                 }
             }
-            if let Some(playback_status) = playback_status {
-                changed_properties.insert("PlaybackStatus".to_owned(), Variant(Box::new(match playback_status {
-                    PlaybackStatus::Playing => "Playing".to_owned(),
-                    PlaybackStatus::Paused => "Paused".to_owned(),
-                    PlaybackStatus::Stopped => "Stopped".to_owned(),
-                })));
+            else {
```
Comment (eladyn) [inline src/dbus_mpris.rs]: You probably want to run `cargo fmt` once to ensure that everything is formatted correctly.

File: src/dbus_mpris.rs [hunk 3]
```diff
@@ -438,42 +439,54 @@ async fn create_dbus_server(
...
-                    let mut m: HashMap<String, Variant<Box<dyn RefArg>>> = HashMap::new();
-                    insert_metadata(&mut m, track);
-
-                    changed_properties.insert("Metadata".to_owned(), Variant(Box::new(m)));
-                } else {
-                    info!("Couldn't fetch metadata from spotify: {:?}", track);
+
+            if last_volume != player_volume {
+                if let Some(player_volume) = player_volume {
+                    changed_properties.insert("Volume".to_owned(), Variant(Box::new(player_volume.to_owned())));
```
Comment (eladyn) [inline src/dbus_mpris.rs]: The `player_volume` here is a `u8` with values that vary from 0 (silent) to 65535 (max volume). (At least my previous observations say so) This is inconsistent with the volume that is returned from the `.Volume` property (which [appears to be](https://github.com/Spotifyd/spotifyd/blob/603bcc4d802d2b8f45ac517c0513eb4096252d7b/src/dbus_mpris.rs#L298-L306) a float from 0.0 to 100.0).

If my assumption that `librespot` just maps the volume to the full available range in a `u8`, is correct (probably should check the librespot code, if this is the case), converting this to a percent value shouldn't be too difficult.

*Note:* In fact, both "scales" don't really comply with the actual MPRIS spec, which [recommends a value from 0.0 to 1.0](https://specifications.freedesktop.org/mpris-spec/latest/Player_Interface.html#Simple-Type:Volume). Don't know however, wether this is something that should be fixed here.

File: src/dbus_mpris.rs [hunk 4]
```diff
@@ -438,42 +439,54 @@ async fn create_dbus_server(
...
             PlayerEvent::Stopped { .. } => {
-                (last_track_id, Some(PlaybackStatus::Stopped))
+                (last_track_id, Some(PlaybackStatus::Stopped), last_volume)
             }
             PlayerEvent::Paused { .. } => {
-                (last_track_id, Some(PlaybackStatus::Paused))
+                (last_track_id, Some(PlaybackStatus::Paused), last_volume)
             }
-            _ => (last_track_id, last_playback_status),
+            _ => (last_track_id, last_playback_status, last_volume),
```
Comment (eladyn) [inline src/dbus_mpris.rs]: We should be able to use `continue` here:

```suggestion
            _ => continue,
```

which might increase the readability. This is, of course, not an issue with your changes, just came to my mind while rereading the code.

**Review by eladyn — CHANGES_REQUESTED** [review]
In general, LGTM and works very well! Only commented on some smaller issues.

File: src/dbus_mpris.rs (hunk 3 above)
Comment (NNEU-1) [reply src/dbus_mpris.rs]: Yep. That's true, I believe a double from 0.0 to 1.0 is the way to go

File: src/dbus_mpris.rs (hunk 4 above)
Comment (NNEU-1) [reply src/dbus_mpris.rs]: That part was unchanged (apart from the addition of last_volume).. 
Let's see what @MDeiml thinks.

File: src/dbus_mpris.rs (hunk 1 above)
Comment (NNEU-1) [reply src/dbus_mpris.rs]: Will do

File: src/dbus_mpris.rs (hunk 4 above)
Comment (MDeiml) [reply src/dbus_mpris.rs]: continue should be fine there, yes

File: src/dbus_mpris.rs (hunk 2 above)
Comment (MDeiml) [reply src/dbus_mpris.rs]: This is still formatted weirdly

**Comment by MDeiml** [comment]
Thanks for implementing this btw.

**Review by MDeiml — COMMENTED** [review]
Is working for me as well. Looks good besides some formatting.

**Comment by NNEU-1** [comment]
Hmm, I don't see any issues with the formatting, at least not with the parts that i've changed..
By the comment it seems that it's the "else" that's bothering you, but I don't see anything wrong with that.

Feel free to change as desired.

**Comment by eladyn** [comment]
@NNEU-1 `cargo fmt -- --check` prints the following for me (on your branch)

```diff
Diff in /home/user/spotifyd/src/dbus_mpris.rs at line 443:
             PlayerEvent::VolumeSet { volume } => {
                 (last_track_id, last_playback_status, Some(volume))
             }
-            PlayerEvent::Playing { track_id, position_ms, .. } => {
+            PlayerEvent::Playing {
+                track_id,
+                position_ms,
+                ..
+            } => {
                 seeked_position = Some(position_ms);
                 (Some(track_id), Some(PlaybackStatus::Playing), last_volume)
             }
Diff in /home/user/spotifyd/src/dbus_mpris.rs at line 457:
         };
 
         // if playback_status, track_id or volume have changed, emit a PropertiesChanged signal
-        if last_playback_status != playback_status || last_track_id != track_id || last_volume != player_volume {
+        if last_playback_status != playback_status
+            || last_track_id != track_id
+            || last_volume != player_volume
+        {
             let mut changed_properties: HashMap<String, Variant<Box<dyn RefArg>>> = HashMap::new();
 
             if last_volume != player_volume {
Diff in /home/user/spotifyd/src/dbus_mpris.rs at line 466:
                     let mut vol_mpris = player_volume as f64;
                     // max. vol = 1.0 according to mpris spec, round to two decimal places
                     vol_mpris = (vol_mpris / 65535.0 * 100.0).round() / 100.0;
-                    changed_properties.insert("Volume".to_owned(), Variant(Box::new(vol_mpris.to_owned())));
+                    changed_properties
+                        .insert("Volume".to_owned(), Variant(Box::new(vol_mpris.to_owned())));
                 }
-            }
-            else {
+            } else {
                 if let Some(track_id) = track_id {
                     let sp = create_spotify_api(&api_token);
                     let track = sp.track(&track_id.to_base62());
Diff in /home/user/spotifyd/src/dbus_mpris.rs at line 483:
                     }
                 }
                 if let Some(playback_status) = playback_status {
-                    changed_properties.insert("PlaybackStatus".to_owned(), Variant(Box::new(match playback_status {
-                        PlaybackStatus::Playing => "Playing".to_owned(),
-                        PlaybackStatus::Paused => "Paused".to_owned(),
-                        PlaybackStatus::Stopped => "Stopped".to_owned(),
-                    })));
+                    changed_properties.insert(
+                        "PlaybackStatus".to_owned(),
+                        Variant(Box::new(match playback_status {
+                            PlaybackStatus::Playing => "Playing".to_owned(),
+                            PlaybackStatus::Paused => "Paused".to_owned(),
+                            PlaybackStatus::Stopped => "Stopped".to_owned(),
+                        })),
+                    );
                 }
             }
```

Some of those things are of course unrelated to your changes. Are you able to reproduce that?

**Comment by NNEU-1** [comment]
> @NNEU-1 `cargo fmt -- --check` prints the following for me (on your branch)
> 
> ```diff
> Diff in /home/user/spotifyd/src/dbus_mpris.rs at line 443:
>              PlayerEvent::VolumeSet { volume } => {
>                  (last_track_id, last_playback_status, Some(volume))
>              }
> -            PlayerEvent::Playing { track_id, position_ms, .. } => {
> +            PlayerEvent::Playing {
> +                track_id,
> +                position_ms,
> +                ..
> +            } => {
>                  seeked_position = Some(position_ms);
>                  (Some(track_id), Some(PlaybackStatus::Playing), last_volume)
>              }
> Diff in /home/user/spotifyd/src/dbus_mpris.rs at line 457:
>          };
>  
>          // if playback_status, track_id or volume have changed, emit a PropertiesChanged signal
> -        if last_playback_status != playback_status || last_track_id != track_id || last_volume != player_volume {
> +        if last_playback_status != playback_status
> +            || last_track_id != track_id
> +            || last_volume != player_volume
> +        {
>              let mut changed_properties: HashMap<String, Variant<Box<dyn RefArg>>> = HashMap::new();
>  
>              if last_volume != player_volume {
> Diff in /home/user/spotifyd/src/dbus_mpris.rs at line 466:
>                      let mut vol_mpris = player_volume as f64;
>                      // max. vol = 1.0 according to mpris spec, round to two decimal places
>                      vol_mpris = (vol_mpris / 65535.0 * 100.0).round() / 100.0;
> -                    changed_properties.insert("Volume".to_owned(), Variant(Box::new(vol_mpris.to_owned())));
> +                    changed_properties
> +                        .insert("Volume".to_owned(), Variant(Box::new(vol_mpris.to_owned())));
>                  }
> -            }
> -            else {
> +            } else {
>                  if let Some(track_id) = track_id {
>                      let sp = create_spotify_api(&api_token);
>                      let track = sp.track(&track_id.to_base62());
> Diff in /home/user/spotifyd/src/dbus_mpris.rs at line 483:
>                      }
>                  }
>                  if let Some(playback_status) = playback_status {
> -                    changed_properties.insert("PlaybackStatus".to_owned(), Variant(Box::new(match playback_status {
> -                        PlaybackStatus::Playing => "Playing".to_owned(),
> -                        PlaybackStatus::Paused => "Paused".to_owned(),
> -                        PlaybackStatus::Stopped => "Stopped".to_owned(),
> -                    })));
> +                    changed_properties.insert(
> +                        "PlaybackStatus".to_owned(),
> +                        Variant(Box::new(match playback_status {
> +                            PlaybackStatus::Playing => "Playing".to_owned(),
> +                            PlaybackStatus::Paused => "Paused".to_owned(),
> +                            PlaybackStatus::Stopped => "Stopped".to_owned(),
> +                        })),
> +                    );
>                  }
>              }
> ```
> 
> Some of those things are of course unrelated to your changes. Are you able to reproduce that?

I haven't tried - I didn't even use cargo for compilation, I just patched the files in my yocto build environment.
So my comments regarding formatting are purely based on my option.

That being said, I don't know the rust formatting conventions. 
I definitively haven't seen "} else {" in C. 

If you want, I can change according to your / cargo fmt's proposal.

**Comment by MDeiml** [comment]
> If you want, I can change according to your / cargo fmt's proposal

That would be nice. I think code submitted to PRs to spotifyd has to be formatted according to `cargo fmt`.

**Comment by NNEU-1** [comment]
Allright, here we go.
That should do :)

**Pull request merged**